src/main/java/com/kryptnostic/datastore/services/CassandraDataManager.java

```diff
@@ -35,95 +49,166 @@
-        return Iterables.transform( entityRows, rs -> RowAdapters.entity( rs, authorizedPropertyTypes ) );
+        Iterable<String> entityIds = getEntityIds( entitySetId, syncIds );
+        Map<String, ResultSetFuture> entityFutures = Maps.toMap( entityIds,
+                entityId -> asyncLoadEntity( entityId, syncIds, authorizedProperties ) );
+        Map<String, ResultSet> entityRows = Maps.transformValues( entityFutures, ResultSetFuture::getUninterruptibly );
+        return Iterables.transform( entityRows.entrySet(),
+                entry -> new Entity(
+                        entry.getKey(),
+                        RowAdapters.entity( entry.getValue(), authorizedPropertyTypes ) ) );
     }
 
-    public Iterable<String> getEntityIds( String entitySetName, Set<UUID> authorizedProperties ) {
+    public Iterable<String> getEntityIds( UUID entitySetId, Set<UUID> syncIds ) {
         BoundStatement boundEntityIdsQuery = entityIdsQuery.bind()
-                .setString( CommonColumns.ENTITY_SETS.cql(), entitySetName )
-                .setSet( CommonColumns.PROPERTY_TYPE.cql(), authorizedProperties );
+                .setSet( CommonColumns.SYNCID.cql(), syncIds )
+                .setUUID( CommonColumns.ENTITY_SET_ID.cql(), entitySetId );
         ResultSet entityIds = session.execute( boundEntityIdsQuery );
         return Iterables.filter( Iterables.transform( entityIds, RowAdapters::entityId ), StringUtils::isNotBlank );
     }
 
-    private ResultSetFuture asyncLoadEntity( String entityId, Set<UUID> authorizedProperties ) {
+    private ResultSetFuture asyncLoadEntity( String entityId, Set<UUID> syncIds, Set<UUID> authorizedProperties ) {
         return session.executeAsync( entitySetQuery.bind()
                 .setString( CommonColumns.ENTITYID.cql(), entityId )
-                .setSet( CommonColumns.PROPERTY_TYPE.cql(), authorizedProperties ) );
+                .setSet( CommonColumns.SYNCID.cql(), syncIds )
+                .setSet( CommonColumns.PROPERTY_TYPE_ID.cql(), authorizedProperties ) );
+    }
+
+    public void createEntityData( CreateEntityRequest req, Set<UUID> authorizedProperties ) {
+        Set<Entity> entities = req.getEntities();
+        UUID syncId = req.getSyncId();
+        UUID entitySetId = req.getEntitySetId();
+
+        List<ResultSetFuture> results = entities.stream().map( entity -> {
+            BatchStatement batch = new BatchStatement();
+
+            batch.add( writeIdLookupQuery.bind().setUUID( CommonColumns.SYNCID.cql(), syncId )
+                    .setUUID( CommonColumns.ENTITY_SET_ID.cql(), entitySetId )
+                    .setString( CommonColumns.ENTITYID.cql(), entity.getId() ) );
+
+            SetMultimap<UUID, Object> propertyValues = entity.getPropertyValues();
+
+            propertyValues.entries().stream().filter( entry -> authorizedProperties.contains( entry.getKey() ) )
```

Comment:
We don't need as much locking here. Just need to make sure we synchronously execute writeIdLookupQuery and then we can do all the individual property's asynchronously. We should wait on all the properties as a whole to finish-- should be pretty with `Futures.allAsList(...)`